kubeflow/trainer/backends/container/utils.py

```diff
@@ -0,0 +1,236 @@
+    if not pkgs:
+        return ""
+
+    index_urls = trainer.pip_index_urls or list(constants.DEFAULT_PIP_INDEX_URLS)
+    main_idx = index_urls[0]
+    extras = " ".join(f"--extra-index-url {u}" for u in index_urls[1:])
+    quoted = " ".join(f'"{p}"' for p in pkgs)
+    return (
+        "PIP_DISABLE_PIP_VERSION_CHECK=1 pip install --no-warn-script-location "
+        f"--index-url {main_idx} {extras} {quoted} && "
+    )
+
+
+def container_status_to_trainjob_status(status: str, exit_code: int) -> str:
+    """
+    Convert container status to TrainJob status.
+
+    Args:
+        status: Container status (e.g., "running", "exited", "created").
+        exit_code: Container exit code.
+
+    Returns:
+        TrainJob status constant.
+    """
+    if status == "running":
+        return constants.TRAINJOB_RUNNING
+    if status == "created":
+        return constants.TRAINJOB_CREATED
+    if status == "exited":
+        # Exit code 0 -> complete, else failed
+        return constants.TRAINJOB_COMPLETE if exit_code == 0 else constants.TRAINJOB_FAILED
+    return constants.UNKNOWN
+
+
+def aggregate_status_from_containers(container_statuses: list[str]) -> str:
+    """
+    Aggregate status from multiple container statuses.
+
+    Args:
+        container_statuses: List of container status strings.
+
+    Returns:
+        Aggregated TrainJob status.
+    """
+    if constants.TRAINJOB_FAILED in container_statuses:
+        return constants.TRAINJOB_FAILED
+    if constants.TRAINJOB_RUNNING in container_statuses:
+        return constants.TRAINJOB_RUNNING
+    if all(s == constants.TRAINJOB_COMPLETE for s in container_statuses if s != UNKNOWN):
+        return constants.TRAINJOB_COMPLETE
+    if any(s == constants.TRAINJOB_CREATED for s in container_statuses):
+        return constants.TRAINJOB_CREATED
+    return UNKNOWN
+
+
+def resolve_image(runtime: types.Runtime) -> str:
+    """
+    Resolve the container image for a runtime from DEFAULT_FRAMEWORK_IMAGES.
+
+    Args:
+        runtime: Runtime object.
+
+    Returns:
+        Container image name.
+
+    Raises:
+        ValueError: If no image is found for the runtime's framework.
+    """
+    framework = runtime.trainer.framework
+    if framework in constants.DEFAULT_FRAMEWORK_IMAGES:
+        return constants.DEFAULT_FRAMEWORK_IMAGES[framework]
+
+    raise ValueError(
+        f"No default image found for framework '{framework}'. "
+        f"Supported frameworks: {list(constants.DEFAULT_FRAMEWORK_IMAGES.keys())}"
+    )
+
+
+def maybe_pull_image(adapter, image: str, pull_policy: str):
+    """
+    Pull image based on pull policy.
+
+    Args:
+        adapter: Container client adapter (DockerClientAdapter or PodmanClientAdapter).
+        image: Container image name.
+        pull_policy: Pull policy ("IfNotPresent", "Always", or "Never").
+
+    Raises:
+        RuntimeError: If image is not found or pull fails.
+    """
+    policy = (pull_policy or "IfNotPresent").lower()
```

Comment:
you don't need this since pull_policy can't be None at this point.
```suggestion
    policy = pull_policy.lower()
```

https://github.com/kubeflow/sdk/blob/9bf4cb7cfd56f5e45c3ceee89023733a6be650db/kubeflow/trainer/backends/container/types.py#L60